kpet/utils.py

```diff
@@ -33,3 +37,106 @@ def get_template_content(filename, dbdir=None):
     template_path = get_template_path(filename, dbdir)
     with open(template_path) as file_handler:
         return file_handler.read()
+
+
+def print2(s):
+    sys.stderr.write(s + '\n')
+
+
+def smart_layout_insert_prefix(f_layout, obj):
+    if obj is None:
+        return None
+    prefix = os.path.dirname(os.path.realpath(f_layout))
+    return '%s/%s' % (prefix, str(obj))
+
+
+def smart_get_layout(f_layout):
+    text = ''
+    with open(f_layout, 'r') as f:
+        text = ''.join(f.readlines())
+
+    layout = json.loads(text, object_pairs_hook=collections.OrderedDict)
+    return layout
+
+
+def smart_get_patterns_bylayout(d_layout, f_layout):
+    l_patterns = []
+    for key in d_layout:
+        val = d_layout[key]
+        if val is None:
+            continue
+
+        s_file = smart_layout_insert_prefix(f_layout, val)
+        text = ''
+        with open(s_file, 'r') as f:
+            text = ''.join(f.readlines())
+
+        p = json.loads(text, object_pairs_hook=collections.OrderedDict)
+        for e in p['patterns']:
+            if e not in l_patterns:
+                l_patterns.append(e)
+
+    return l_patterns
+
+
+def smart_get_src_files(l_patch):
+    """ Get source files according to list of patch files
+
+        N.B. Typical source files shown in patch file look like:
+             diff --git a/fs/btrfs/ctree.h b/fs/btrfs/ctree.h
+             diff --git a/fs/btrfs/file.c b/fs/btrfs/file.c
+             ...
+    """
+    p = re.compile(r'^diff --git a/.* b/.*')
+
+    l_src = []
+    for f_patch in l_patch:
+        with open(f_patch, 'r') as f:
+            while True:
+                line = f.readline()
+                if len(line) == 0:
+                    break
+
+                if p.match(line.strip()) is None:
+                    continue
+
+                line = line.rstrip()
+                l_line = line.split(' ')
+                srcfile = l_line[-1].replace('b/', '')
+                if srcfile not in l_src:
+                    l_src.append(srcfile)
+
+    return l_src
+
+
+def smart_get_test_cases(l_src, f_layout):
+    # get layout
+    d_layout = smart_get_layout(f_layout)
+    if d_layout is None:
```

Comment:
The same comment than here https://github.com/RH-FMK/kpet/pull/13/files#diff-d628452b9af638b1cd6e05b21bf0bb5dR48

Reply:
Updated.